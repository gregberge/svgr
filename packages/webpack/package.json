{
  "name": "@svgr/webpack",
  "description": "SVGR webpack loader.",
  "version": "3.1.0",
  "main": "lib/index.js",
  "repository": "git@github.com:smooth-code/svgr.git",
  "author": "Greg Berg√© <berge.greg@gmail.com>",
  "keywords": [
    "svgr",
    "svg",
    "react",
    "webpack",
    "webpack-loader"
  ],
  "engines": {
    "node": ">=8"
  },
  "license": "MIT",
  "scripts": {
    "prebuild": "rm -rf lib/",
    "build": "babel --config-file ../../babel.config.js -d lib --ignore \"**/*.test.js\" src",
    "prepublishOnly": "yarn run build"
  },
  "dependencies": {
    "@babel/core": "^7.1.2",
    "@babel/plugin-transform-react-constant-elements": "^7.0.0",
    "@babel/preset-env": "^7.1.0",
    "@babel/preset-react": "^7.0.0",
    "@svgr/core": "^3.1.0",
    "loader-utils": "^1.1.0"
  },
  "devDependencies": {
    "babel-loader": "^8.0.4",
    "memory-fs": "^0.4.1",
    "url-loader": "^1.1.2",
    "webpack": "^4.20.2"
  }
}
