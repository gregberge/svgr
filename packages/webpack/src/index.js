import { getOptions } from 'loader-utils'
import { transform as babelTransform } from '@babel/core'
import convert from '@svgr/core'

function svgrLoader(source) {
  const callback = this.async()
  const { babel = true, ...options } = getOptions(this) || {}

  const readSvg = () =>
    new Promise((resolve, reject) => {
      this.fs.readFile(this.resourcePath, (err, result) => {
        if (err) reject(err)
        resolve(result)
      })
    })

  const exportMatches = source
    .toString('utf-8')
    .match(/^module.exports\s*=\s*(.*)/)
  const previousExport = exportMatches ? exportMatches[1] : null

  const pBabelTransform = async jsCode =>
    new Promise((resolve, reject) => {
      babelTransform(
        jsCode,
        {
          babelrc: false,
          presets: [
            '@babel/preset-react',
            ['@babel/preset-env', { modules: false }],
          ],
          plugins: [
            '@babel/plugin-transform-react-constant-elements',
            '@babel/plugin-proposal-object-rest-spread',
          ],
        },
        (err, result) => {
          if (err) reject(err)
          else resolve(result.code)
        },
      )
    })

  readSvg()
    .then(svg => convert(svg, options, { webpack: { previousExport } }))
    .then(jsCode => (babel ? pBabelTransform(jsCode) : jsCode))
    .then(result => callback(null, result))
    .catch(err => callback(err))
}

export default svgrLoader
