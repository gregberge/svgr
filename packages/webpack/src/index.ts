import { callbackify } from 'util'
import { transformAsync, createConfigItem } from '@babel/core'
import { transform, Config, State } from '@svgr/core'
import { normalize } from 'path'
import svgo from '@svgr/plugin-svgo'
import jsx from '@svgr/plugin-jsx'
// @ts-ignore
import presetReact from '@babel/preset-react'
// @ts-ignore
import presetEnv from '@babel/preset-env'
// @ts-ignore
import presetTS from '@babel/preset-typescript'
// @ts-ignore
import pluginTransformReactConstantElements from '@babel/plugin-transform-react-constant-elements'
import type * as webpack from 'webpack'

const babelOptions = {
  babelrc: false,
  configFile: false,
  presets: [
    createConfigItem(presetReact, { type: 'preset' }),
    createConfigItem([presetEnv, { modules: false }], { type: 'preset' }),
  ],
  plugins: [createConfigItem(pluginTransformReactConstantElements)],
}

const typeScriptBabelOptions = {
  ...babelOptions,
  presets: [
    ...babelOptions.presets,
    createConfigItem(
      [presetTS, { allowNamespaces: true, allExtensions: true, isTSX: true }],
      { type: 'preset' },
    ),
  ],
}

interface LoaderOptions extends Config {
  babel?: boolean
}

const tranformSvg = callbackify(
  async (contents: string, options: LoaderOptions, state: Partial<State>) => {
    const { babel = true, ...config } = options
    const jsCode = await transform(contents, config, state)
    if (!babel) return jsCode
    const result = await transformAsync(
      jsCode,
      options.typescript ? typeScriptBabelOptions : babelOptions,
    )
    if (!result?.code) {
      throw new Error(`Error while transforming using Babel`)
    }
    return result.code
  },
)

function svgrLoader(
  this: webpack.LoaderContext<LoaderOptions>,
  contents: string,
): void {
  this.cacheable && this.cacheable()
  const callback = this.async()

  const options = this.getOptions()

  const previousExport = (() => {
    if (contents.startsWith('export ')) return contents
    const exportMatches = contents.match(/^module.exports\s*=\s*(.*)/)
    return exportMatches ? `export default ${exportMatches[1]}` : null
  })()

  const state = {
    caller: {
      name: '@svgr/webpack',
      previousExport,
      defaultPlugins: [svgo, jsx],
    },
    filePath: normalize(this.resourcePath),
  }

  if (!previousExport) {
    tranformSvg(contents, options, state, callback)
  } else {
    this.fs.readFile(this.resourcePath, (err, result) => {
      if (err) {
        callback(err)
        return
      }
      tranformSvg(String(result), options, state, (err, content) => {
        if (err) {
          callback(err)
          return
        }
        callback(null, content)
      })
    })
  }
}

export default svgrLoader
