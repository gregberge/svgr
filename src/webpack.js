import { getOptions } from 'loader-utils'
import { transform as babelTransform } from '@babel/core'
import convert from './'

function svgrLoader(source) {
  const callback = this.async()
  const { babel = true, ...options } = getOptions(this) || {}

  const readSvg = () =>
    new Promise((resolve, reject) => {
      this.fs.readFile(this.resourcePath, (err, result) => {
        if (err) reject(err)
        resolve(result)
      })
    })

  const exportMatches = source.match(/^module.exports\s*=\s*(.*)/)
  const previousExport = exportMatches ? exportMatches[1] : null

  const pBabelTransform = async jsCode =>
    new Promise((resolve, reject) => {
      babelTransform(
        jsCode,
        {
          babelrc: false,
          presets: [
            '@babel/preset-react',
            ['@babel/preset-env', { modules: false }],
          ],
          plugins: [
            '@babel/plugin-transform-react-constant-elements',
            '@babel/plugin-proposal-object-rest-spread',
          ],
        },
        (err, result) => {
          if (err) reject(err)
          else resolve(result.code)
        },
      )
    })

  readSvg()
    .then(svg => convert(svg, options, { webpack: { previousExport } }))
    .then(jsCode => (babel ? pBabelTransform(jsCode) : jsCode))
    .then(result => callback(null, result))
    .catch(err => callback(err))
}

function svgrLoaderWithWarning(source) {
  /* eslint-disable no-console */
  console.warn(
    'Using "svgr/lib/webpack" is deprecated and will be removed in v2. Please use "svgr/webpack" instead.',
  )
  /* eslint-enable no-console */
  return svgrLoader.call(this, source)
}

export { svgrLoader }
export default svgrLoaderWithWarning
