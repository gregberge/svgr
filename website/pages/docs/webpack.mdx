---
section: Usage
title: Webpack
order: 20
---

# Webpack

SVGR provides an official [webpack.js](https://webpack.js.org/) loader to import SVG as React components.

<carbon-ad />

## Install

```bash
npm install --save-dev @svgr/webpack
# or use yarn
yarn add --dev @svgr/webpack
```

## Usage

**webpack.config.js**

```js
module.exports = {
  module: {
    rules: [
      {
        test: /\.svg$/i,
        issuer: /\.[jt]sx?$/,
        use: ['@svgr/webpack'],
      },
    ],
  },
}
```

**Your code**

```js
import Star from './star.svg'

const Example = () => (
  <div>
    <Star />
  </div>
)
```

## Options

SVGR let you specify options in a runtime config file like `svgr.config.js` or directly in the loader:

**webpack.config.js**

```js
module.exports = {
  module: {
    rules: [
      {
        test: /\.svg$/i,
        issuer: /\.[jt]sx?$/,
        use: [{ loader: '@svgr/webpack', options: { icon: true } }],
      },
    ],
  },
}
```

[SVGR options reference](https://react-svgr.com/docs/options/) describes all options available.

## Use SVGR and asset SVG in the same project

You may be interested to use some SVG as an asset (url) and other SVG as a React component. The easiest way to do it is to use a `resourceQuery` for one of the two type.

**webpack.config.js**

```js
module.exports = {
  module: {
    rules: [
      {
        test: /\.svg$/i,
        type: 'asset',
        resourceQuery: /url/, // *.svg?url
      },
      {
        test: /\.svg$/i,
        issuer: /\.[jt]sx?$/,
        resourceQuery: { not: [/url/] }, // exclude react component if *.svg?url
        use: ['@svgr/webpack'],
      },
    ],
  },
}
```

**Your code**

```js
import svg from './assets/file.svg?url'
import Svg from './assets/file.svg'

const App = () => {
  return (
    <div>
      <img src={svg} width="200" height="200" />
      <Svg width="200" height="200" viewBox="0 0 3500 3500" />
    </div>
  )
}
```

## Use SVG in CSS files

The `issuer: /\.[jt]sx?$/` option ensures that SVGR will only apply if the SVG is imported from a JavaScript or a TypeScript file. It allows you to safely import SVG into your `.css` or `.scss` without any issue.

```css
.example {
  background-image: url(./assets/file.svg);
}
```

## Use with `url-loader` or `file-loader`

> `url-loader` and `file-loader` are deprecated over [Asset Modules](https://webpack.js.org/guides/asset-modules/) in webpack v5. It is widely encouraged to use `resourceQuery` method described before.

SVGR can be used with [`url-loader`](https://github.com/webpack-contrib/url-loader) or [`file-loader`](https://github.com/webpack-contrib/file-loader).

**webpack.config.js**

```js
module.exports = {
  module: {
    rules: [
      {
        test: /\.svg$/i,
        issuer: /\.[jt]sx?$/,
        use: ['@svgr/webpack', 'url-loader'],
      },
    ],
  },
}
```

**Your code**

```js
import starUrl, { ReactComponent as Star } from './star.svg'

const App = () => (
  <div>
    <img src={starUrl} alt="star" />
    <Star />
  </div>
)
```

The named export defaults to `ReactComponent` and can be customized with the `namedExport` option.

Please note that by default, `@svgr/webpack` will try to export the React Component via default export if there is no other loader handling svg files with default export. When there is already any other loader using default export for svg files, `@svgr/webpack` will always export the React component via named export.

If you prefer named export in any case, you may set the `exportType` option to `named`.

## Use your own Babel configuration

By default, `@svgr/webpack` includes a `babel-loader` with [an optimized configuration](https://github.com/gregberge/svgr/blob/main/packages/webpack/src/index.ts). In some case you may want to apply a custom one (if you are using Preact for an example). You can turn off Babel transformation by specifying `babel: false` in options.

```js
// Example using preact
{
  test: /\.svg$/,
  use: [
    {
      loader: 'babel-loader',
      options: {
        presets: ['preact', 'env'],
      },
    },
    {
      loader: '@svgr/webpack',
      options: { babel: false },
    }
  ],
}
```
