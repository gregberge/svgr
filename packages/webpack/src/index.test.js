import path from 'path'
import webpack from 'webpack'
import MemoryFs from 'memory-fs'

function compile(rules) {
  const compiler = webpack({
    mode: 'development',
    context: __dirname,
    entry: './__fixtures__/icon.svg',
    output: {
      path: __dirname,
      filename: 'bundle.js',
    },
    module: { rules },
  })

  compiler.outputFileSystem = new MemoryFs()

  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) reject(err)

      resolve(
        stats
          .toJson()
          .modules.find(({ name }) => name === './__fixtures__/icon.svg')
          .source,
      )
    })
  })
}

describe('webpack loader', () => {
  it(
    'should convert file',
    async () => {
      const source = await compile([
        {
          test: /\.svg$/,
          use: [
            {
              loader: path.resolve(__dirname, './index.js'),
              options: {
                expandProps: false,
              },
            },
          ],
        },
      ])

      expect(source).toMatchSnapshot()
    },
    15000,
  )

  it(
    'should support url-loader',
    async () => {
      const source = await compile([
        {
          test: /\.svg$/,
          use: [
            {
              loader: path.resolve(__dirname, './index.js'),
              options: {
                expandProps: false,
              },
            },
            'url-loader',
          ],
        },
      ])

      expect(source).toMatchSnapshot()
    },
    15000,
  )

  it(
    'should convert file (babel: false)',
    async () => {
      const source = await compile([
        {
          test: /\.svg$/,
          use: [
            {
              loader: 'babel-loader',
              options: {
                babelrc: false,
                presets: ['@babel/preset-react'],
              },
            },
            {
              loader: path.resolve(__dirname, './index.js'),
              options: {
                babel: false,
                expandProps: false,
              },
            },
          ],
        },
      ])

      expect(source).toMatchSnapshot()
    },
    15000,
  )
})
