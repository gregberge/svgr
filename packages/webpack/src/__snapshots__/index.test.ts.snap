// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack loader supports url-loader 1`] = `
"var _svg;

import * as React from \\"react\\";

var SvgIcon = function SvgIcon() {
  return _svg || (_svg = /*#__PURE__*/React.createElement(\\"svg\\", {
    width: 88,
    height: 88,
    xmlns: \\"http://www.w3.org/2000/svg\\"
  }, /*#__PURE__*/React.createElement(\\"g\\", {
    stroke: \\"#063855\\",
    strokeWidth: 2,
    fill: \\"none\\",
    fillRule: \\"evenodd\\",
    strokeLinecap: \\"square\\"
  }, /*#__PURE__*/React.createElement(\\"path\\", {
    d: \\"M51 37 37 51M51 51 37 37\\"
  }))));
};

export { SvgIcon as ReactComponent };
export default \\"data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4KPHN2ZyB3aWR0aD0iODhweCIgaGVpZ2h0PSI4OHB4IiB2aWV3Qm94PSIwIDAgODggODgiIHZlcnNpb249IjEuMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayI+CiAgICA8IS0tIEdlbmVyYXRvcjogU2tldGNoIDQ2LjIgKDQ0NDk2KSAtIGh0dHA6Ly93d3cuYm9oZW1pYW5jb2RpbmcuY29tL3NrZXRjaCAtLT4KICAgIDx0aXRsZT5EaXNtaXNzPC90aXRsZT4KICAgIDxkZXNjPkNyZWF0ZWQgd2l0aCBTa2V0Y2guPC9kZXNjPgogICAgPGRlZnM+PC9kZWZzPgogICAgPGcgaWQ9IkJsb2NrcyIgc3Ryb2tlPSJub25lIiBzdHJva2Utd2lkdGg9IjEiIGZpbGw9Im5vbmUiIGZpbGwtcnVsZT0iZXZlbm9kZCIgc3Ryb2tlLWxpbmVjYXA9InNxdWFyZSI+CiAgICAgICAgPGcgaWQ9IkRpc21pc3MiIHN0cm9rZT0iIzA2Mzg1NSIgc3Ryb2tlLXdpZHRoPSIyIj4KICAgICAgICAgICAgPHBhdGggZD0iTTUxLDM3IEwzNyw1MSIgaWQ9IlNoYXBlIj48L3BhdGg+CiAgICAgICAgICAgIDxwYXRoIGQ9Ik01MSw1MSBMMzcsMzciIGlkPSJTaGFwZSI+PC9wYXRoPgogICAgICAgIDwvZz4KICAgIDwvZz4KPC9zdmc+Cg==\\";"
`;

exports[`webpack loader transforms file (babel: false) 1`] = `
"import * as React from \\"react\\";

const SvgIcon = () => /*#__PURE__*/React.createElement(\\"svg\\", {
  width: 88,
  height: 88,
  xmlns: \\"http://www.w3.org/2000/svg\\"
}, /*#__PURE__*/React.createElement(\\"g\\", {
  stroke: \\"#063855\\",
  strokeWidth: 2,
  fill: \\"none\\",
  fillRule: \\"evenodd\\",
  strokeLinecap: \\"square\\"
}, /*#__PURE__*/React.createElement(\\"path\\", {
  d: \\"M51 37 37 51M51 51 37 37\\"
})));

export default SvgIcon;"
`;

exports[`webpack loader transforms file (typescript: true) 1`] = `
"var _g;

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import * as React from \\"react\\";

var SvgIcon = function SvgIcon(props) {
  return /*#__PURE__*/React.createElement(\\"svg\\", _extends({
    width: 88,
    height: 88,
    xmlns: \\"http://www.w3.org/2000/svg\\"
  }, props), _g || (_g = /*#__PURE__*/React.createElement(\\"g\\", {
    stroke: \\"#063855\\",
    strokeWidth: 2,
    fill: \\"none\\",
    fillRule: \\"evenodd\\",
    strokeLinecap: \\"square\\"
  }, /*#__PURE__*/React.createElement(\\"path\\", {
    d: \\"M51 37 37 51M51 51 37 37\\"
  }))));
};

export default SvgIcon;"
`;

exports[`webpack loader transforms file 1`] = `
"var _svg;

import * as React from \\"react\\";

var SvgIcon = function SvgIcon() {
  return _svg || (_svg = /*#__PURE__*/React.createElement(\\"svg\\", {
    width: 88,
    height: 88,
    xmlns: \\"http://www.w3.org/2000/svg\\"
  }, /*#__PURE__*/React.createElement(\\"g\\", {
    stroke: \\"#063855\\",
    strokeWidth: 2,
    fill: \\"none\\",
    fillRule: \\"evenodd\\",
    strokeLinecap: \\"square\\"
  }, /*#__PURE__*/React.createElement(\\"path\\", {
    d: \\"M51 37 37 51M51 51 37 37\\"
  }))));
};

export default SvgIcon;"
`;

exports[`webpack loader transforms file 2`] = `
"var _svg;

import * as React from \\"react\\";

var SvgIcon = function SvgIcon() {
  return _svg || (_svg = /*#__PURE__*/React.createElement(\\"svg\\", {
    width: 88,
    height: 88,
    xmlns: \\"http://www.w3.org/2000/svg\\"
  }, /*#__PURE__*/React.createElement(\\"g\\", {
    stroke: \\"#063855\\",
    strokeWidth: 2,
    fill: \\"none\\",
    fillRule: \\"evenodd\\",
    strokeLinecap: \\"square\\"
  }, /*#__PURE__*/React.createElement(\\"path\\", {
    d: \\"M51 37 37 51M51 51 37 37\\"
  }))));
};

export default SvgIcon;"
`;
